### PR TITLE
feat(ui): add keyboard shortcuts

### LINKED ISSUE #756: Add keyboard shortcuts
Global:
`⌘`+`T` or `⌘`+`P` or `⌘`+`F` or `S` (maybe we should decide on one) - open search
`⌘`+`,` opens Settings

Profile page (org and user)
`⌘`+`N` to create new project

I'll add more if I think of them.

### PR DESCRIPTION
**Pull request by juliendonck**
Adds keyboard shortcuts:
// To open help => OS modifier key + /
// To open settings => OS modifier key + ,
// To open search => OS modifier key + p
// To open design system => OS modifier key + d
// To create a new project => OS modifier key + n

Also uses `cmd` key on mac and `ctrl` on other OS's for all shortcuts.

closes #756

### CODE REVIEW

**Comment by juliendonck** [comment]
@xla can you test if this works on linux? Should work with the `ctrl`-key.

File: ui/Hotkeys.svelte [hunk 1]
```diff
@@ -11,16 +10,37 @@
...
+      toggle(path.help());
+    }
+
+    // To open settings => OS modifier key + ,
+    if (OSKey && event.code === "Comma") {
+      toggle(path.settings());
+    }
+
+    // To open search => OS modifier key + p
+    if (OSKey && event.code === "Comma") {
```
Comment (rudolfs) [inline ui/Hotkeys.svelte:32]:
```suggestion
    if (OSKey && event.code === "KeyP") {
```

File: ui/Screen/Help.svelte [hunk 2]
```diff
@@ -1,40 +1,58 @@
 <script>
   import ModalLayout from "../DesignSystem/Component/ModalLayout.svelte";
+
+  const OSKey = navigator.platform.includes("Mac") ? "cmd" : "ctrl";
```
Comment (rudolfs) [inline ui/Screen/Help.svelte:4]:
```suggestion
  const OSKey = navigator.platform.includes("Mac") ? "⌘" : "ctrl";
```

Comment (juliendonck) [reply ui/Screen/Help.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/759/commits/5a04df924950e2f19780497bd4417679715bd42c

File: ui/Hotkeys.svelte (hunk 1 above)
Comment (juliendonck) [reply ui/Hotkeys.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/759/commits/5a04df924950e2f19780497bd4417679715bd42c

**Comment by rudolfs** [comment]
I wonder if there's a nice way to signal the user that keyboard shortcuts are available in the app?

**Comment by juliendonck** [comment]
> I wonder if there's a nice way to signal the user that keyboard shortcuts are available in the app?

@rudolfs I was wondering about that as well. Maybe we can put it in the settings in some way. @brandonhaslegs what do you think?

**Comment by brandonhaslegs** [comment]
Normally they're in tooltips.
Slack:
image

Figma:
image

**Comment by rudolfs** [comment]
Hm, maybe we can stick them in the tooltips and get rid of the help screen?

**Comment by juliendonck** [comment]
> Normally they're in tooltips.
> Slack:
> image
> 
> Figma:
> image

@rudolfs @brandonhaslegs what would be the icon for `ctrl`?

File: ui/Screen/Help.svelte [hunk 3]
```diff
@@ -1,40 +1,58 @@
 <script>
   import ModalLayout from "../DesignSystem/Component/ModalLayout.svelte";
+
+  const OSKey = navigator.platform.includes("Mac") ? "⌘" : "ctrl";
```
Comment (rudolfs) [inline ui/Screen/Help.svelte:4]: Actually this should be `modifierKey`.

Comment (juliendonck) [reply ui/Screen/Help.svelte:4]: What do you mean?

Comment (juliendonck) [reply ui/Screen/Help.svelte:4]: I have both a mac and windows at home and all the actions I do with the `cmd` key are mapped to `ctrl` on windows.

Comment (rudolfs) [reply ui/Screen/Help.svelte:4]: I meant the name of the variable.

Comment (rudolfs) [reply ui/Screen/Help.svelte:4]: So `OSKey` -> `modifierKey`.

**Comment by rudolfs** [comment]
> @rudolfs @brandonhaslegs what would be the icon for ctrl?

I don't think it has an icon, it should be just "Ctrl": https://en.wikipedia.org/wiki/Control_key.

File: ui/Screen/Help.svelte (hunk 3 above)
Comment (juliendonck) [reply ui/Screen/Help.svelte:4]: Oh you mean the variable?

**Comment by juliendonck** [comment]
> > @rudolfs @brandonhaslegs what would be the icon for ctrl?
> 
> I don't think it has an icon, it should be just "Ctrl": https://en.wikipedia.org/wiki/Control_key.
> Or maybe `^`?

@rudolfs yeah but I think that's a mac thing. Not sure windows people get it...

**Comment by rudolfs** [comment]
> @rudolfs yeah but I think that's a mac thing. Not sure windows people get it...

I think <kbd>Ctrl</kbd> is the safest option.

**Comment by xla** [comment]
Shouldn't it show the key depending on the OS, i.e. <kbd>Cmd</kbd> for mac and <kbd>Ctrl</kbd> for other platforms?

**Comment by juliendonck** [comment]
> Shouldn't it show the key depending on the OS, i.e. Cmd for mac and Ctrl for other platforms?

@xla jup that's what I've implemented. @rudolfs and I are just talking about how to display `ctrl` in the app.

File: ui/Hotkeys.svelte [hunk 4]
```diff
@@ -11,16 +10,37 @@
...
+
+    // To open design system => OS modifier key + d
+    if (OSKey && event.code === "KeyD") {
       toggle(path.designSystemGuide());
     }
 
-    if (event.shiftKey && event.code === "Slash") {
-      toggle(path.help());
+    // To create a new project => OS modifier key + n
+    if (OSKey && event.code === "KeyN") {
```
Comment (xla) [inline ui/Hotkeys.svelte:42]: Shouldn't this be sensitive to the context? Feels a bit odd to jump to the project creation when you browse org members for example.

File: ui/Hotkeys.svelte [hunk 5]
```diff
@@ -11,16 +10,37 @@
   };
 
   const onKeydown = event => {
+    const OSKey = navigator.platform.includes("Mac")
```
Comment (xla) [inline ui/Hotkeys.svelte:13]:
```suggestion
    const osKey = navigator.platform.includes("Mac")
```

Convention in js/ts/svelte is camelCase for var/let/const names.

File: ui/Screen/Help.svelte (hunk 3 above)
Comment (xla) [reply ui/Screen/Help.svelte:4]: We do the same check in `Hotkeys.svelte`, maybe worth moving to lib code.

Comment (xla) [reply ui/Screen/Help.svelte:4]: +1 on naming it `modifierKey` in both cases.

File: ui/Hotkeys.svelte (hunk 4 above)
Comment (juliendonck) [reply ui/Hotkeys.svelte:42]: Yeah that's how @brandonhaslegs had specced it in the issue. But I was thinking that right now the only thing you can really create is a new project. So I don't think it's a big deal to have this happen from anywhere, I can see in the future this opens a modal where you can choose what new thing you want to create:
- project
- org
- transfer
- etc

Comment (xla) [reply ui/Hotkeys.svelte:42]: Understood.

File: ui/Hotkeys.svelte [hunk 6]
```diff
@@ -1,6 +1,5 @@
 <script>
   import { location, pop, push } from "svelte-spa-router";
-
```
[the remark this answers is not in the document]
Comment (juliendonck) [reply ui/Hotkeys.svelte]: https://github.com/radicle-dev/radicle-upstream/commit/79c719e7030813f8769a00892116d4b1c31397b3

File: ui/Hotkeys.svelte (hunk 5 above)
Comment (juliendonck) [reply ui/Hotkeys.svelte]: https://github.com/radicle-dev/radicle-upstream/commit/79c719e7030813f8769a00892116d4b1c31397b3

File: ui/Screen/Help.svelte (hunk 3 above)
Comment (juliendonck) [reply ui/Screen/Help.svelte]: https://github.com/radicle-dev/radicle-upstream/commit/79c719e7030813f8769a00892116d4b1c31397b3

**Comment by juliendonck** [comment]
@xla @rudolfs @brandonhaslegs decided to keep it simple for now and create a link in the settings to our keyboard shortcuts. It didn't make sense to put them in the tooltips for now since there are only 2 places where you'd do it (`settings` and `new project`) and it didn't make sense to put a tooltip on the `new project` button every time you hovered it. Should do for now.

Screenshot 2020-08-07 at 12 38 57

File: ui/src/settings.ts [hunk 7]
```diff
@@ -58,3 +58,7 @@ export const themeOptions: Option[] = [
     value: Theme.Dark,
   },
 ];
+
+// gives back the OS you're using in hotkeys.svelte & shortcuts.svelte
+
```
Comment (rudolfs) [inline ui/src/settings.ts:63]:
```suggestion
```

**Comment by juliendonck** [comment]
> Triggering a new keyboard shortcut that opens a modal should close any previous modals right? With these shortcuts I can myself get into a situation where I have many modals open on top of each other and closing reveals all the previous ones I had open. For example, try making a new project, then opening the keyboard shortcuts modal, then new project, then shortcuts modal, etc, etc… it opens endless modals on top of one another. Maybe this can be reported in another issue though.
> 
> Excited to have keyboard shortcuts!

@brandonhaslegs our modal world is a separate concern that we won't be focusing on for now. It's a long story...

**Pull request merged**